lib/services/local/access_list.go

```diff
@@ -136,19 +137,37 @@ func (a *AccessListService) GetAccessListsToReview(ctx context.Context) ([]*acce
 
 // UpsertAccessList creates or updates an access list resource.
 func (a *AccessListService) UpsertAccessList(ctx context.Context, accessList *accesslist.AccessList) (*accesslist.AccessList, error) {
-	err := a.service.RunWhileLocked(ctx, lockName(accessList.GetName()), accessListLockTTL, func(ctx context.Context, _ backend.Backend) error {
-		ownerMap := make(map[string]struct{}, len(accessList.Spec.Owners))
-		for _, owner := range accessList.Spec.Owners {
-			if _, ok := ownerMap[owner.Name]; ok {
-				return trace.AlreadyExists("owner %s already exists in the owner list", owner.Name)
+	upsertWithLockFn := func() error {
+		return a.service.RunWhileLocked(ctx, lockName(accessList.GetName()), accessListLockTTL, func(ctx context.Context, _ backend.Backend) error {
+			ownerMap := make(map[string]struct{}, len(accessList.Spec.Owners))
+			for _, owner := range accessList.Spec.Owners {
+				if _, ok := ownerMap[owner.Name]; ok {
+					return trace.AlreadyExists("owner %s already exists in the owner list", owner.Name)
+				}
+				ownerMap[owner.Name] = struct{}{}
 			}
-			ownerMap[owner.Name] = struct{}{}
+			return trace.Wrap(a.service.UpsertResource(ctx, accessList))
+		})
+	}
+
+	var completeUpsertFn func() error
+	if feature := modules.GetModules().Features(); !feature.IdentityGovernance {
+		completeUpsertFn = func() error {
+			return a.service.RunWhileLocked(ctx, "createAccessListLimitLock", accessListLockTTL, func(ctx context.Context, _ backend.Backend) error {
+				if err := a.VerifyAccessListCreateLimit(ctx, accessList.GetName()); err != nil {
+					return trace.Wrap(err)
+				}
+				return trace.Wrap(upsertWithLockFn())
+			})
 		}
-		return trace.Wrap(a.service.UpsertResource(ctx, accessList))
-	})
-	if err != nil {
+	} else {
+		completeUpsertFn = upsertWithLockFn
```

Comment:
Maybe that's only me, but this looks misleading. as this function is not reused anywhere I think that a "direct" call would be easier to understand

```go
func (a *AccessListService) UpsertAccessList(ctx context.Context, accessList *accesslist.AccessList) (*accesslist.AccessList, error) {
	var err error
	if feature := modules.GetModules().Features(); !feature.IdentityGovernance {
		err = func() error {
			return a.service.RunWhileLocked(ctx, "createAccessListLimitLock", accessListLockTTL, func(ctx context.Context, _ backend.Backend) error {
				if err := a.VerifyAccessListCreateLimit(ctx, accessList.GetName()); err != nil {
					return trace.Wrap(err)
				}
				return trace.Wrap(upsertWithLockFn())
			})
		}()
	} else {
		err = func() error {
			return a.service.RunWhileLocked(ctx, lockName(accessList.GetName()), accessListLockTTL, func(ctx context.Context, _ backend.Backend) error {
				ownerMap := make(map[string]struct{}, len(accessList.Spec.Owners))
				for _, owner := range accessList.Spec.Owners {
					if _, ok := ownerMap[owner.Name]; ok {
						return trace.AlreadyExists("owner %s already exists in the owner list", owner.Name)
					}
					ownerMap[owner.Name] = struct{}{}
				}
				return trace.Wrap(a.service.UpsertResource(ctx, accessList))
		})()
	}

	if err != nil {
		return nil, trace.Wrap(err)
	}

	return accessList, nil
}
```

Second more important question, why do we have two separate implementations here? Looks like on checks for the usage limit, but the other one upserts with a lock?

Reply:
hrm, sorry i'm not sure i understand 😅 . i followed the pattern used with [DeviceTrust](https://github.com/gravitational/teleport.e/blob/0652dbadfd4cefc80bb3c44c456bef7f8f8cd5f3/lib/devicetrust/storage/storage.go#L1174)

The reason why `upsertWithLock` was created was to avoid `copy pastying` the same logic to create an access list in both conditions where
   1) condition one: we need to verify limit first, before creating. And we need to grab a lock for `verifying` named `createAccessListLimitLock` so no two verification can happen at same time, and then upon success allow creating with its own lock so no two access list is updated at same time. 
   2) condition two: we don't need to verify, thus i didn't need to wrap the create logic with the verify lock

i hope that made sense (that was my undesrtanding when i read the device trust logic)